### ISSUE TITLE
robot.api.TestData methods takes file path instead of file object type

### ISSUE THREAD

**Issue by 1dziubo1** [robotframework/robotframework#2852]
Examples why methods shouldn't do that is plenty across whole internet. Mine, for example, lies in remote svn repository, so no path to file on my OS. I could use svn methods to cat contents of file, or even save it to memory, but now i have to save file to some tmp path (that is susceptible t all IOErrors), just to se method and remove. With over 16GB repository size, full of those tests, this significantly slows down my work!

**Comment by chriscallan** [comment]
I can see that it would be really beneficial to be able to refer to any file anywhere and have RF take care of getting there.  It would seem to be a complex task of figuring out what the endpoint is, loading necessary libraries to work with the file, and run the tests.  Tests may require that other files be created and doing so on an Svn repo wouldn't be appropriate (in most cases).  Not to mention that this all needs to be done for Windows, Linux, and Mac.
Local IOErrors are easy to troubleshoot, but when you start working with IOErrors across a network connection there are a lot of other things that can go wrong.
Likely, it would be easier to pull down copies of the repos you need to test and work with them locally, which negates all the above because there's only a filesystem driver involved and the OS is already handling that for us.
It would be a "nice" addition for RF (I've never needed this functionality), but I think the complexity of the above would probably stop most folks in their tracks.  If you have a solution for this, put together a PR and see if it can be included.

**Comment by pekkaklarck** [comment]
I totally agree functions accepting paths to files to process should generally also accept file and file-like objects. Many of the Robot's APIs behave like this already, and others can be enhanced when it makes sense. With the `TestData` API there are few problems that make this a bit complicated:

- This factory method works both with file and directory paths. File objects could only be used instead of file paths. 
- The code handling file paths uses `os.path.dirname` to get the parent directory of the parsed file. I don't remember exactly where it is used, but I suspect it's used at least with imports (e.g. `Resource    foo.robot` assumes `foo.robot` to be relative to that directory) and when resolving the special `${CURDIR}` variable. With file objects this needed to be handled somehow differently.

In other words, I consider this a useful feature but it may be somewhat hard to implement. Are you @1dziubo1 interested to look at implementing this?

@chriscallan, I don't think the point was to be able to refer to any file anywhere, but instead to pass files as open file or file-like objects. These objects can then work as facades to data somewhere else.

**Comment by 1dziubo1** [comment]
I see your points, both for and against. Rather than rewriting current solution (for me code is not well-readable, still, it works), I'd suggest class, that can parse robot file to object static-like. I think, that will solve problem of backwards compability, legacy and generally, touching that code. In addition, RF can have nice tool to verify  test suite syntax (without dependencies).

In my case, this is most suites, that shouldn't be in that repo in first place :(
(printing errors in Windows, crashed Windows terminal, first time in my life :P )

How I see this, would be object like:
class RFTestSuite(object):
    def __init__(self, source_file, suite name):
          self.suite_name = suite name
          self.test_cases = get_test_cases() #list of RFTestCase objects
          self.keywords = get_keywords() # list of RFKeyword objects
          self.documentation = ...
          self.tags = ...
          self.libraries = ....

I think you get the point :) I see huge advantages in having RF suite file object-like representation:
- verification (e.g. precommit hook fail if documentation is None)
- syntax check and repair (e.g. there is an error, if someone has '***Settings***' instead '*** Settings ***'), fixing code to CamelCase or replacing tabs with spaces
- anti patterns detection (throw a stone, who has never met testers copy-paste technique in case of keywords, with that we would be able if keyword with same structure is already written)
- RECREATING SUITE OUT OF OBJECT (static only), that would have consistent scheme (e.g. same structure, with documentation first, libraries second and so on; keywords in alphabetical order - human friendly template)

Okay, this is definitely out of scope of this issue. Designing such classes wouldn't be problem, however methods would require adapting your current solutions (I could do that on strings, totally reinventing wheel here, not to mention working on bare strings would be... challenging to put it lightly)

**Comment by chriscallan** [comment]
Very good distinction about the open file objects vs. paths to files.
Thanks for catching that.



Chris Callan

*Senior QA Engineer*

www.Base2S.com

*Cell:* (206) 383-5829

*"Real difficulties can be overcome; it is only the imaginary ones that are
unconquerable."*

*--Theodore Vail (industrialist)*



*From:* Pekka Klärck <notifications@github.com>
*Sent:* Tuesday, June 5, 2018 11:14 PM
*To:* robotframework/robotframework <robotframework@noreply.github.com>
*Cc:* chriscallan <chris.callan@base2s.com>; Mention <
mention@noreply.github.com>
*Subject:* Re: [robotframework/robotframework] robot.api.TestData methods
takes file path instead of file object type (#2852)



I totally agree functions accepting paths to files to process should
generally also accept file and file-like objects. Many of the Robot's APIs
behave like this already, and others can be enhanced when it makes sense.
With the TestData API there are few problems that make this a bit
complicated:

   - This factory method works both with file and directory paths. File
   objects could only be used instead of file paths.
   - The code handling file paths uses os.path.dirname to get the parent
   directory of the parsed file. I don't remember exactly where it is used,
   but I suspect it's used at least with imports (e.g. Resource foo.robot
   assumes foo.robot to be relative to that directory) and when resolving
   the special ${CURDIR} variable. With file objects this needed to be
   handled somehow differently.

In other words, I consider this a useful feature but it may be somewhat
hard to implement. Are you @1dziubo1 <https://github.com/1dziubo1>
interested to look at implementing this?

@chriscallan <https://github.com/chriscallan>, I don't think the point was
to be able to refer to any file anywhere, but instead to pass files as open
file or file-like objects. These objects can then work as facades to data
somewhere else.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/robotframework/robotframework/issues/2852#issuecomment-394952193>,
or mute the thread

**Comment by pekkaklarck** [comment]
Old parsing modules are going to be replaced in RF 3.2 by #3076. Issue related to the old modules aren't relevant anymore.
